Support for restrictions

jesper-friis opened the pull request:
# Description
Added support for adding and finding restrictions.

Also fixed a bug in the Literal.n3() method. 

**Question**: The `utils.tfilter()` function was added to support finding restrictions, but in the end it was never used. Should we remove it? Or is it nice to have for storage-efficient filtering of the output of triple generators?

**Note**: This PR builds on top of #187, so merge #187 first.

## Type of change
- [x] Bug fix and code cleanup
- [x] New feature
- [ ] Documentation update
- [x] Testing

## Checklist for the reviewer
This checklist should be used as a help for the reviewer.
- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?
- [ ] Is the code properly tested?

francescalb commented on tripper/triplestore.py:
```diff
@@ -673,6 +680,153 @@ def prefix_iri(self, iri: str, require_prefixed: bool = False):
+        target: "Union[str, Literal]",
+        type: "TypeLiteral['some', 'only', 'exactly', 'min', 'max', 'value']",
+        cardinality: "Optional[int]" = None,
+        hashlength: int = 16,
```
Should not this also be typeset to optional?

jesper-friis replied:
I don't think so. `hashlength` must always be an integer. The "Optional" type annotation indicates that you can provide None as a default value. That is not allowed for `hashlength`.

francescalb reviewed (changes requested):
Looks  good, but codecov points out a few additional things that shold be tested.

jesper-friis commented:
> Looks good, but codecov points out a few additional things that shold be tested.

Added important missing test for finding value restrictions with literals. Reveiled a bug that is now fixed!

I have not added tests for utils.tfilter(). I ended up not using this function. Should we keep it?

francescalb commented:
> > Looks good, but codecov points out a few additional things that shold be tested.
> 
> Added important missing test for finding value restrictions with literals. Reveiled a bug that is now fixed!
> 
> I have not added tests for corner cases with utils.tfilter() since I ended up not using this function. Should we keep it?

I like the concpet of the function though. Is it difficult to add the tests?

jesper-friis commented:
> > > Looks good, but codecov points out a few additional things that shold be tested.
> > 
> > 
> > Added important missing test for finding value restrictions with literals. Reveiled a bug that is now fixed!
> > I have not added tests for corner cases with utils.tfilter() since I ended up not using this function. Should we keep it?
> 
> I like the concpet of the function though. Is it difficult to add the tests?

Added three additional tests for tfilter. Now it is fully covered...

francescalb reviewed (approved):
Nice
